src/gepa/core/engine.py

```diff
@@ -117,129 +128,172 @@ def _run_full_eval_and_add(
-                                )
-                                # Skip reflective this iteration (old behavior)
-                                continue
-                    # Old behavior: regardless of whether we attempted, clear the flag before reflective
-                    self.merge_proposer.last_iter_found_new_program = False
-
-                # 2) Reflective mutation proposer
-                proposal = self.reflective_proposer.propose(state)
-                if proposal is None:
-                    self.logger.log(f"Iteration {state.i + 1}: Reflective mutation did not propose a new candidate")
-                    continue
-
-                # Acceptance: require strict improvement on subsample
-                old_sum = sum(proposal.subsample_scores_before or [])
-                new_sum = sum(proposal.subsample_scores_after or [])
-                if new_sum <= old_sum:
-                    self.logger.log(f"Iteration {state.i + 1}: New subsample score is not better, skipping")
-                    continue
-
-                # Accept: full eval + add
-                self._run_full_eval_and_add(
-                    new_program=proposal.candidate, state=state, parent_program_idx=proposal.parent_program_ids
-                )
-
-                # Schedule merge attempts like original behavior
-                if self.merge_proposer is not None:
-                    self.merge_proposer.last_iter_found_new_program = True
-                    if self.merge_proposer.total_merges_tested < self.merge_proposer.max_merge_invocations:
-                        self.merge_proposer.merges_due += 1
-
-            except Exception as e:
-                self.logger.log(f"Iteration {state.i + 1}: Exception during optimization: {e}")
-                self.logger.log(traceback.format_exc())
-                if self.raise_on_exception:
-                    raise e
-                else:
-                    continue
-
-        # Close progress bar if it exists
-        if self.display_progress_bar:
-            progress_bar.close()
-
-        state.save(self.run_dir)
-        return state
+                signal.signal(sig, handler)
+            except (OSError, ValueError):
+                pass
+
+    @contextmanager
+    def _signal_handler_context(self):
+        """Context manager for signal handling setup and restoration."""
+        if self.enable_signal_handling:
+            self._setup_signal_handlers()
+        try:
+            yield
+        finally:
+            if self.enable_signal_handling:
+                self._restore_signal_handlers()
+
+    def _should_stop(self, state: GEPAState) -> bool:
+        """Check if the optimization should stop."""
+        if self._stop_requested:
+            return True
+        if self.stop_callback and self.stop_callback():
+            return True
+        if state.total_num_evals >= self.max_metric_calls:
```

Comment:
This should be a stopper right? Implement it as a stopper, and in engine, build a composite stopper, which includes any stop callback provided by the user + max_metric_calls stopper.